File: examples/python/agg-table/handlers/start_hand.py
"""Handler for StartHand command."""

import sys
from pathlib import Path
from datetime import datetime, timezone
import uuid

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

# Add paths for imports
sys.path.insert(0, str(Path(__file__).parent.parent.parent / "angzarr"))
sys.path.insert(0, str(Path(__file__).parent.parent))

from angzarr_client.proto.angzarr import types_pb2 as types
from angzarr_client.errors import CommandRejectedError
from angzarr_client.proto.examples import table_pb2 as table

from .state import TableState


def handle_start_hand(
    command_book: types.CommandBook,
    command_any: Any,
    state: TableState,
    seq: int,
) -> types.EventBook:
    """Handle StartHand command."""
    if not state.exists():
        raise CommandRejectedError("Table does not exist")

    if state.status == "in_hand":
        raise CommandRejectedError("Hand already in progress")

    # Need at least 2 active players
    active_count = state.active_player_count()
    if active_count < 2:
        raise CommandRejectedError("Not enough players to start hand")

    # Generate hand root (deterministic from table + hand number)
    hand_number = state.hand_count + 1
    hand_uuid = uuid.uuid5(
        uuid.NAMESPACE_DNS,
        f"angzarr.poker.hand.{state.table_id}.{hand_number}",
    )
    hand_root = hand_uuid.bytes

    # Advance dealer button
    dealer_position = state.next_dealer_position()

    # Get active player positions for blind calculation
    active_positions = sorted(
        pos for pos, seat in state.seats.items() if not seat.is_sitting_out
    )

    # Find small blind and big blind positions
    dealer_idx = 0
    for i, pos in enumerate(active_positions):
        if pos == dealer_position:
            dealer_idx = i
            break

    if len(active_positions) == 2:
        # Heads up: dealer is small blind
        sb_position = active_positions[dealer_idx]
        bb_position = active_positions[(dealer_idx + 1) % 2]
    else:
        sb_position = active_positions[(dealer_idx + 1) % len(active_positions)]
        bb_position = active_positions[(dealer_idx + 2) % len(active_positions)]

    # Build active players list
    active_players = []
    for pos in active_positions:
        seat = state.seats[pos]
        active_players.append(
            table.SeatSnapshot(
                position=pos,
                player_root=seat.player_root,
                stack=seat.stack,
            )
        )

    event = table.HandStarted(
        hand_root=hand_root,
        hand_number=hand_number,
        dealer_position=dealer_position,
        small_blind_position=sb_position,
        big_blind_position=bb_position,
        game_variant=state.game_variant,
        small_blind=state.small_blind,
        big_blind=state.big_blind,
        started_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )
    event.active_players.extend(active_players)

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.poker/")

    return types.EventBook(
        cover=command_book.cover,
        pages=[
            types.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(
                    seconds=int(datetime.now(timezone.utc).timestamp())
                ),
            )
        ],
    )

File: examples/go/agg-table/handlers/start_hand.go
package handlers

import (
	"crypto/sha256"
	"encoding/binary"
	"time"

	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// HandleStartHand handles the StartHand command.
func HandleStartHand(
	commandBook *pb.CommandBook,
	commandAny *anypb.Any,
	state TableState,
	seq uint32,
) (*pb.EventBook, error) {
	if !state.Exists() {
		return nil, angzarr.NewCommandRejectedError("Table does not exist")
	}

	// Just verify it's the right command type
	var cmd examples.StartHand
	if err := proto.Unmarshal(commandAny.Value, &cmd); err != nil {
		return nil, err
	}

	// Check current state
	if state.Status == "in_hand" {
		return nil, angzarr.NewCommandRejectedError("Hand already in progress")
	}

	// Need at least 2 active players
	if state.ActivePlayerCount() < 2 {
		return nil, angzarr.NewCommandRejectedError("Not enough players to start hand")
	}

	// Generate hand root from table root + hand number
	handNumber := state.HandCount + 1
	handRoot := generateHandRoot(commandBook.Cover.Root.Value, handNumber)

	// Advance dealer position to next active player
	dealerPosition := advanceToNextActive(state.DealerPosition, state)
	smallBlindPosition := advanceToNextActive(dealerPosition, state)
	bigBlindPosition := advanceToNextActive(smallBlindPosition, state)

	// Build active player snapshots
	var activePlayers []*examples.SeatSnapshot
	for _, seat := range state.Seats {
		if !seat.IsSittingOut {
			activePlayers = append(activePlayers, &examples.SeatSnapshot{
				Position:   seat.Position,
				PlayerRoot: seat.PlayerRoot,
				Stack:      seat.Stack,
			})
		}
	}

	event := &examples.HandStarted{
		HandRoot:           handRoot,
		HandNumber:         handNumber,
		DealerPosition:     dealerPosition,
		SmallBlindPosition: smallBlindPosition,
		BigBlindPosition:   bigBlindPosition,
		ActivePlayers:      activePlayers,
		GameVariant:        state.GameVariant,
		SmallBlind:         state.SmallBlind,
		BigBlind:           state.BigBlind,
		StartedAt:          timestamppb.New(time.Now()),
	}

	eventAny, err := anypb.New(event)
	if err != nil {
		return nil, err
	}
	eventAny.TypeUrl = "type.poker/examples.HandStarted"

	return &pb.EventBook{
		Cover: commandBook.Cover,
		Pages: []*pb.EventPage{
			{
				Sequence:  &pb.EventPage_Num{Num: seq},
				Event:     eventAny,
				CreatedAt: timestamppb.New(time.Now()),
			},
		},
	}, nil
}

// generateHandRoot creates a deterministic hand root from table root and hand number.
func generateHandRoot(tableRoot []byte, handNumber int64) []byte {
	h := sha256.New()
	h.Write(tableRoot)
	buf := make([]byte, 8)
	binary.BigEndian.PutUint64(buf, uint64(handNumber))
	h.Write(buf)
	return h.Sum(nil)
}

// advanceToNextActive finds the next active (non-sitting-out) player position.
func advanceToNextActive(currentPos int32, state TableState) int32 {
	maxPlayers := state.MaxPlayers
	for i := int32(1); i <= maxPlayers; i++ {
		nextPos := (currentPos + i) % maxPlayers
		if seat, exists := state.Seats[nextPos]; exists && !seat.IsSittingOut {
			return nextPos
		}
	}
	return currentPos // Shouldn't happen if we have active players
}
